### PR TITLE
eiffelstudio: fix build on Linux and Monterey

### PR DESCRIPTION
**Pull request by danielnachun**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/eiffelstudio.rb [hunk 1]
```diff
@@ -17,20 +17,32 @@ class Eiffelstudio < Formula
...
-    (bin/"vision2_demo").write_env_script(prefix/"vision2_demo/spec/macosx-x86-64/bin/vision2_demo", env)
+    env = { ISE_EIFFEL: prefix, ISE_PLATFORM: "#{os}-x86-64" }
+    (bin/"ec").write_env_script(prefix/"studio/spec/#{os}-x86-64/bin/ec", env)
+    (bin/"ecb").write_env_script(prefix/"studio/spec/#{os}-x86-64/bin/ecb", env)
+    (bin/"estudio").write_env_script(prefix/"studio/spec/#{os}-x86-64/bin/estudio", env)
+    (bin/"finish_freezing").write_env_script(prefix/"studio/spec/#{os}-x86-64/bin/finish_freezing", env)
+    (bin/"compile_all").write_env_script(prefix/"tools/spec/#{os}-x86-64/bin/compile_all", env)
+    (bin/"iron").write_env_script(prefix/"tools/spec/#{os}-x86-64/bin/iron", env)
+    (bin/"syntax_updater").write_env_script(prefix/"tools/spec/#{os}-x86-64/bin/syntax_updater", env)
+    (bin/"vision2_demo").write_env_script(prefix/"vision2_demo/spec/#{os}-x86-64/bin/vision2_demo", env)
```
Comment (carlocab) [inline Formula/eiffelstudio.rb:45]:
```suggestion
    eiffel_env = { ISE_EIFFEL: prefix, ISE_PLATFORM: "#{os}-x86-64" }
    {
      studio:       ["ec", "ecb", "estudio", "finish_freezing"],
      tools:        ["compile_all", "iron", "syntax_updater"],
      vision2_demo: ["vision2_demo"],
    }.each do |subdir, targets|
      targets.each do |target|
        (bin/target).write_env_script prefix/subdir/"spec/#{os}-x86-64"/target, eiffel_env
      end
    end
```

File: Formula/eiffelstudio.rb [hunk 2]
```diff
@@ -17,20 +17,32 @@ class Eiffelstudio < Formula
...
+        "-flat_namespace -undefined suppress",
+        "-undefined dynamic_lookup"
+      system "tar", "cjf", "c.tar.bz2", "C"
+    end
+
+    os = OS.mac? ? "macosx" : OS.kernel_name.downcase
+    system "./compile_exes", "#{os}-x86-64"
+    system "./make_images", "#{os}-x86-64"
     prefix.install Dir["Eiffel_19.05/*"]
     bin.mkpath
```
Comment (carlocab) [inline Formula/eiffelstudio.rb:36]:
```suggestion
```
I think this isn't needed.

File: Formula/eiffelstudio.rb [hunk 3]
```diff
@@ -17,20 +17,32 @@ class Eiffelstudio < Formula
...
+    if OS.mac?
+      system "tar", "xf", "c.tar.bz2"
+      inreplace "C/CONFIGS/macosx-x86-64",
+        "-flat_namespace -undefined suppress",
+        "-undefined dynamic_lookup"
+      system "tar", "cjf", "c.tar.bz2", "C"
+    end
+
+    os = OS.mac? ? "macosx" : OS.kernel_name.downcase
+    system "./compile_exes", "#{os}-x86-64"
```
Comment (carlocab) [inline Formula/eiffelstudio.rb:33]: `x86-64` everywhere seems weird.

File: Formula/eiffelstudio.rb [hunk 4]
```diff
@@ -17,20 +17,32 @@ class Eiffelstudio < Formula
...
+
   def install
-    system "./compile_exes", "macosx-x86-64"
-    system "./make_images", "macosx-x86-64"
+    # Fix flat namespace usage in C shared library.
+    if OS.mac?
+      system "tar", "xf", "c.tar.bz2"
+      inreplace "C/CONFIGS/macosx-x86-64",
+        "-flat_namespace -undefined suppress",
+        "-undefined dynamic_lookup"
```
Comment (carlocab) [inline Formula/eiffelstudio.rb:28]:
```suggestion
      inreplace "C/CONFIGS/macosx-x86-64",
                "-flat_namespace -undefined suppress",
                "-undefined dynamic_lookup"
```
Though I'd put this in one line if it meets the line limit.

File: Formula/eiffelstudio.rb (hunk 3 above)
Comment (danielnachun) [reply Formula/eiffelstudio.rb:33]: Should we maybe just use something like `os_tag = "#{os}-x86-64"`?  And then use that variable everywhere instead?

Comment (carlocab) [reply Formula/eiffelstudio.rb:33]: We could, but I guess my point was that it's `x86-64` even on ARM.

Comment (carlocab) [reply Formula/eiffelstudio.rb:33]: `brew audit` will fail if we've been doing that.

Comment (danielnachun) [reply Formula/eiffelstudio.rb:33]: Hmm, so should we be changing anything?.  It's certainly odd and confusing but I'm not sure anything can be done right now.

File: Formula/eiffelstudio.rb [hunk 5]
```diff
@@ -17,20 +17,31 @@ class Eiffelstudio < Formula
...
-    (bin/"syntax_updater").write_env_script(prefix/"tools/spec/macosx-x86-64/bin/syntax_updater", env)
-    (bin/"vision2_demo").write_env_script(prefix/"vision2_demo/spec/macosx-x86-64/bin/vision2_demo", env)
+    eiffel_env = { ISE_EIFFEL: prefix, ISE_PLATFORM: os_tag }
+    {
+      studio:       %w[ec ecb estudio finish_freezing],
+      tools:        %w[compile_all iron syntax_updater],
+      vision2_demo: %w[vision2_demo],
+    }.each do |subdir, targets|
+      targets.each do |target|
+        (bin/target).write_env_script prefix/subdir/"spec/#{os_tag}"/target, eiffel_env
```
Comment (carlocab) [inline Formula/eiffelstudio.rb:42]:
```suggestion
        (bin/target).write_env_script prefix/subdir/"spec/#{os_tag}/bin"/target, eiffel_env
```

**Comment by danielnachun** [comment]
@carlocab the current failure seems to be because in:
```
    {
      studio:       %w[ec ecb estudio finish_freezing],
      tools:        %w[compile_all iron syntax_updater],
      vision2_demo: %w[vision2_demo],
    }.each do |subdir, targets|
      targets.each do |target|
        (bin/target).write_env_script prefix/subdir/"spec"/os_tag/"bin"/target, eiffel_env
      end
    end
```
`subdir` is a Symbol rather than a String.  I was able to fix this by converting `subdir` to a String but I'm not sure what the cleanest way is to do this should be.  I just changed it to:
```
    {
      studio:       %w[ec ecb estudio finish_freezing],
      tools:        %w[compile_all iron syntax_updater],
      vision2_demo: %w[vision2_demo],
    }.each do |subdir, targets|
      subdir_s = subdir.to_s
      targets.each do |target|
        (bin/target).write_env_script prefix/subdir_s/"spec"/os_tag/"bin"/target, eiffel_env
      end
    end
```
but there may be something better we can do instead.

File: Formula/eiffelstudio.rb [hunk 6]
```diff
@@ -17,20 +17,31 @@ class Eiffelstudio < Formula
...
-    (bin/"syntax_updater").write_env_script(prefix/"tools/spec/macosx-x86-64/bin/syntax_updater", env)
-    (bin/"vision2_demo").write_env_script(prefix/"vision2_demo/spec/macosx-x86-64/bin/vision2_demo", env)
+    eiffel_env = { ISE_EIFFEL: prefix, ISE_PLATFORM: os_tag }
+    {
+      studio:       %w[ec ecb estudio finish_freezing],
+      tools:        %w[compile_all iron syntax_updater],
+      vision2_demo: %w[vision2_demo],
+    }.each do |subdir, targets|
+      targets.each do |target|
+        (bin/target).write_env_script prefix/subdir/"spec"/os_tag/"bin"/target, eiffel_env
```
Comment (carlocab) [inline Formula/eiffelstudio.rb:42]:
```suggestion
        (bin/target).write_env_script prefix/subdir.to_s/"spec"/os_tag/"bin"/target, eiffel_env
```

**Comment by danielnachun** [comment]
I think this is ready now.

**Review by carlocab — APPROVED** [review]

**Pull request closed without merging**